topic/topic.go

```diff
@@ -0,0 +1,205 @@
+		SetNumberOfShards(t.NumberOfShards).
+		SetConsumerServices(css), nil
+}
+
+func (t *topic) Name() string {
+	return t.name
+}
+
+func (t *topic) SetName(value string) Topic {
+	newt := *t
+	newt.name = value
+	return &newt
+}
+
+func (t *topic) NumberOfShards() uint32 {
+	return t.numOfShards
+}
+
+func (t *topic) SetNumberOfShards(value uint32) Topic {
+	newt := *t
+	newt.numOfShards = value
+	return &newt
+}
+
+func (t *topic) ConsumerServices() []ConsumerService {
+	return t.consumerServices
+}
+
+func (t *topic) SetConsumerServices(value []ConsumerService) Topic {
+	newt := *t
+	newt.consumerServices = value
+	return &newt
+}
+
+// ToProto creates proto from a topic.
+func ToProto(t Topic) (*topicpb.Topic, error) {
+	css := t.ConsumerServices()
+	csspb := make([]*topicpb.ConsumerService, len(css))
+	for i, cs := range css {
+		cspb, err := ConsumerServiceToProto(cs)
+		if err != nil {
+			return nil, err
+		}
+		csspb[i] = cspb
+	}
+	return &topicpb.Topic{
+		Name:             t.Name(),
+		NumberOfShards:   t.NumberOfShards(),
+		ConsumerServices: csspb,
+	}, nil
+}
+
+type consumerService struct {
+	sid services.ServiceID
+	ct  ConsumptionType
+}
+
+// NewConsumerService creates a ConsumerService.
+func NewConsumerService() ConsumerService {
+	return new(consumerService)
+}
+
+// NewConsumerServiceFromProto creates a ConsumerService from a proto.
+func NewConsumerServiceFromProto(cs *topicpb.ConsumerService) (ConsumerService, error) {
+	ct, err := NewConsumptionTypeFromProto(cs.ConsumptionType)
+	if err != nil {
+		return nil, err
+	}
+	return NewConsumerService().
+		SetServiceID(NewServiceIDFromProto(cs.ServiceId)).
+		SetConsumptionType(ct), nil
+}
+
+// ConsumerServiceToProto creates proto from a ConsumerService.
+func ConsumerServiceToProto(cs ConsumerService) (*topicpb.ConsumerService, error) {
+	ct, err := ConsumptionTypeToProto(cs.ConsumptionType())
+	if err != nil {
+		return nil, err
+	}
+	return &topicpb.ConsumerService{
+		ConsumptionType: ct,
+		ServiceId:       ServiceIDToProto(cs.ServiceID()),
+	}, nil
+}
+
+func (cs *consumerService) ServiceID() services.ServiceID {
+	return cs.sid
+}
+
+func (cs *consumerService) SetServiceID(value services.ServiceID) ConsumerService {
+	newcs := *cs
+	newcs.sid = value
+	return &newcs
+}
+
+func (cs *consumerService) ConsumptionType() ConsumptionType {
+	return cs.ct
+}
+
+func (cs *consumerService) SetConsumptionType(value ConsumptionType) ConsumerService {
+	newcs := *cs
+	newcs.ct = value
+	return &newcs
+}
+
+func (cs *consumerService) String() string {
+	return fmt.Sprintf("{service: %s, consumption type: %s}", cs.sid.String(), cs.ct.String())
+}
+
+// NewConsumptionTypeFromProto creates ConsumptionType from a proto.
+func NewConsumptionTypeFromProto(ct topicpb.ConsumptionType) (ConsumptionType, error) {
+	switch ct {
+	case topicpb.ConsumptionType_SHARED:
+		return Shared, nil
+	case topicpb.ConsumptionType_REPLICATED:
+		return Replicated, nil
+	}
+	var c ConsumptionType
```

Comment:
Should we define an `Unknown` consumption type? Otherwise this will just be Shared by default.

Reply:
sure